File: services/app/src/interfaces/Manga.ts
export default interface Manga {
    data: Datum[];
  }
  
  export interface Datum {
    id:            string;
    type:          string;
    attributes:    DatumAttributes;
    relationships: Relationship[];
  }
  
  export interface DatumAttributes {
    title:                          TitleClass;
    altTitles:                      AltTitle[];
    description:                    TitleClass;
    isLocked:                       boolean;
    links:                          Links;
    originalLanguage:               string;
    lastVolume:                     string;
    lastChapter:                    string;
    publicationDemographic:         null | string;
    status:                         string;
    year:                           number | null;
    contentRating:                  string;
    tags:                           Tag[];
    state:                          string;
    chapterNumbersResetOnNewVolume: boolean;
    createdAt:                      Date;
    updatedAt:                      Date;
    version:                        number;
    availableTranslatedLanguages:   string[];
    latestUploadedChapter:          null | string;
  }
  
  export interface AltTitle {
    ja?:      string;
    en?:      string;
    "ja-ro"?: string;
  }
  
  export interface TitleClass {
    en?: string;
  }
  
  export interface Links {
    al?:  string;
    mu?:  string;
    amz?: string;
    raw?: string;
  }
  
  export interface Tag {
    id:            string;
    type:          Type;
    attributes:    TagAttributes;
    relationships: any[];
  }
  
  export interface TagAttributes {
    name:        TitleClass;
    description: PurpleDescription;
    group:       Group;
    version:     number;
  }
  
  export interface PurpleDescription {
  }
  
  export enum Group {
    Format = "format",
    Genre = "genre",
    Theme = "theme",
  }
  
  export enum Type {
    Tag = "tag",
  }
  
  export interface Relationship {
    id:          string;
    type:        string;
    attributes?: RelationshipAttributes;
  }
  
  export interface RelationshipAttributes {
    description: string;
    volume:      string;
    fileName:    string;
    locale:      string;
    createdAt:   Date;
    updatedAt:   Date;
    version:     number;
  }
  

File: services/api/v1/src/models/manga.rs
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Manga {
    id: String,
    #[serde(rename = "type")]
    kind: String,
    attributes: MangaAttributes,
    relationships: Vec<MangaRelationship>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
struct MangaTagAttributes {
    name: MultiLanguageSupport,
    description: MultiLanguageSupport,
    group: String,
    version: i32,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
struct MangaTag {
    id: String,
    #[serde(rename = "type")]
    kind: String,
    attributes: MangaTagAttributes,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
struct MangaRelationshipAttributes {
    description: String,
    volume: String,
    #[serde(rename = "fileName")]
    file_name: String,
    locale: String,
    #[serde(rename = "createdAt")]
    created_at: String,
    #[serde(rename = "updatedAt")]
    updated_at: String,
    version: i32,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
struct MangaRelationship {
    id: String,
    #[serde(rename = "type")]
    kind: String,
    attributes: Option<MangaRelationshipAttributes>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
struct MultiLanguageSupport {
    en: Option<String>,
    ja: Option<String>,
    #[serde(rename = "ja-ro")]
    ja_ro: Option<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
struct MangaAttributes {
    title: MultiLanguageSupport,
    #[serde(rename = "altTitles")]
    alt_itles: Vec<MultiLanguageSupport>,
    description: MultiLanguageSupport,
    #[serde(rename = "isLocked")]
    is_locked: bool,
    #[serde(rename = "originalLanguage")]
    original_language: String,
    status: String,
    year: i32,
    #[serde(rename = "contentRating")]
    content_rating: String,
    tags: Vec<MangaTag>,
    state: String,
    version: i32,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct MangaResponse {
    result: String,
    response: String,
    data: Vec<Manga>,
    total: i32,
    offset: i32,
    limit: i32,
}
